### PR TITLE
charts/gateway - redis integration

### PR DESCRIPTION
**Pull request by Gazza7205**
**Description of the change**
- Updated image
  - Updated to Gateway 11.0.00_CR1
    - this will cause a restart if you are not overriding the default image
- Redis Integration
  - [Redis Configuration](#redis-configuration) options for the Gateway (future use)
  - Added Redis SubChart
- Ingress
  - Backend service is now more configurable allowing the management service to be exposed via ingress controller
    - ***this should only be done in environments where the ingress controller does not have a Public Address***
    - ingress.rules[n]backend can be set to "management"
- Restart on config change
  - A new flag has been added to facilitate auto redeploy of Gateways when there is a config change
  - Applies to the default config map only
    - does not include config.cwp, config.listenPorts or the Gateway Secret
- MySQL subChart updated

**Benefits**
- Updated Gateway Image to 11.0.00_CR1
- Redis integration options
- [Remove the need to use rollout restart when updating configmaps](https://github.com/CAAPIM/apim-charts/pull/261)
- Route to the Management service via Ingress Controller

**Drawbacks**
- This update will cause a rolling restart of the Gateway Deployment (new image)

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[charts/gateway]`)
- [X] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

### CODE REVIEW

File: charts/gateway/README.md [hunk 1]
```diff
@@ -678,6 +697,35 @@ ingress:
...
+# com.l7tech.server.extension.sharedCounterProvider=redis
+# com.l7tech.server.extension.sharedRateLimiterProvider=redis
+```
+
+| Parameter                        | Description                               | Default                                                      |
+| -----------------------------    | -----------------------------------       | -----------------------------------------------------------  |
+| `config.redis.enabled`          | Enable redis configuration | `false`  |
+| `config.redis.existingConfigSecret`          | Use an existing config secret - must contain a key called redis.properties | `false`  |
+| `config.redis.subChart.enabled`          | Deploy the redis subChart | `true`  |
+| `config.redis.groupName`          | Redis Group name | ``  |
```
Comment (jagsgill) [inline charts/gateway/README.md]: The Gateway uses default group name `l7GW` if the user doesn't define their own

File: charts/gateway/values.yaml [hunk 2]
```diff
@@ -214,8 +214,11 @@ config:
     javax.net.ssl.trustStoreType=jks
     # Period of time before the Gateway removes inactive nodes.
     com.l7tech.server.clusterStaleNodeCleanupTimeoutSeconds=86400
+    # Redis configuration - uncomment to use Redis
```
Comment (jagsgill) [inline charts/gateway/values.yaml:217]: Does uncommenting these lines override the defaults on line 206-208? Or will those need to be commented/removed as well?

File: charts/gateway/README.md (hunk 1 above)
Comment (gvermeulen7205) [reply charts/gateway/README.md:713]: I'll add that as a default

Comment (Gazza7205) [reply charts/gateway/README.md]: I'll add that as a default

File: charts/gateway/values.yaml (hunk 2 above)
Comment (Gazza7205) [reply charts/gateway/values.yaml:217]: It does override them in my tests - that's a good point. I'll add a note in the readme to comment those out.

Comment (Gazza7205) [reply charts/gateway/values.yaml:217]: line 208 is not overridden - should that be redis too?

Comment (jagsgill) [reply charts/gateway/values.yaml:217]: you're correct, not 208 (sharedClusterInfoProvider is not currently implemented with redis). Only 206 & 207

**Review by DamandeepToor — APPROVED** [review]
LGTM for restartOnConfigChange flag

**Review by jagsgill — APPROVED** [review]
Reviewed the Redis related changes, LGTM. Internal validation was completed already

**Review by burbanski — COMMENTED** [review]
LTGM

File: charts/gateway/README.md [hunk 3]
```diff
@@ -678,6 +697,41 @@ ingress:
...
+```
+# com.l7tech.server.extension.sharedKeyValueStoreProvider=redis
+# com.l7tech.server.extension.sharedCounterProvider=redis
+# com.l7tech.server.extension.sharedRateLimiterProvider=redis
+```
+
+| Parameter                        | Description                               | Default                                                      |
+| -----------------------------    | -----------------------------------       | -----------------------------------------------------------  |
+| `config.redis.enabled`          | Enable redis configuration | `false`  |
+| `config.redis.existingConfigSecret`          | Use an existing config secret - must contain a key called redis.properties | `false`  |
```
Comment (ghost) [inline charts/gateway/README.md]: This field should be the name of the secret instead of true/false

File: charts/gateway/README.md [hunk 4]
```diff
@@ -678,6 +697,41 @@ ingress:
...
+| `config.redis.auth.password.encoded`          | Password is encoded | `false`  |
+| `config.redis.auth.password.value`          | Redis password | `mypassword`  |
+| `config.redis.sentinel.enabled`                | Enable sentinel configuration   | `true` |
+| `config.redis.sentinel.masterSet`          | Redis Master set | `mymaster`  |
+| `config.redis.sentinel.nodes`          | Array of sentinel nodes and ports | `[]`  |
+| `config.redis.standalone.host`                | Redis host if sentinel is not enabled   | `redis-standalone` |
+| `config.redis.standalone.port`                | Redis port if sentinel is not enabled   | `6379` |
+| `config.redis.tls.enabled`    | Enable SSL/TLS              | `false` |
+| `config.redis.tls.existingSecret`    | Use an existing secret - must contain a key called tls.crt        | `` |
+| `config.redis.tls.verifyPeer`    | Verify Peer             | `false` |
```
Comment (ghost) [inline charts/gateway/README.md]: We might want default verifyPeer to true

File: charts/gateway/production-values.yaml [hunk 5]
```diff
@@ -354,6 +357,49 @@ config:
...
+      # If the subChart is not enabled sentinel nodes need to be set
+      nodes:
+      - <host>:<port>
+      - <host>:<port>
+      - <host>:<port>
+    standalone: {}
+    #  host: redis-standalone
+    #  port: 6379
+    tls:
+      enabled: true
```
Comment (ghost) [inline charts/gateway/production-values.yaml:412]: We should update here to be consistent of the default value in Readme

File: charts/gateway/README.md (hunk 4 above)
Comment (Gazza7205) [reply charts/gateway/README.md]: verifyPeer now defaults to true

File: charts/gateway/production-values.yaml (hunk 5 above)
Comment (Gazza7205) [reply charts/gateway/production-values.yaml:412]: Updated

**Comment by Gazza7205** [comment]
Closing - new PR here. https://github.com/CAAPIM/apim-charts/pull/283

**Pull request closed without merging**